src/Swift.Bindings/src/Demangler/MatchRule.cs

```diff
@@ -0,0 +1,133 @@
+// Licensed under the MIT License.
+
+namespace BindingsGeneration.Demangling;
+/// <summary>
+/// MatchRule represents a tree reducing rule for a demangled Swift symbol.
+/// Given a tree of nodes, a match rule contains conditions for the match,
+/// which include:
+/// - If and how the node content should be matched
+/// - If the number of children should match
+/// - Rules to run on the children
+/// - One or more NodeKinds to match
+/// If a match occurs, a reducer function can be run on the node.
+/// </summary>
+internal class MatchRule {
+    /// <summary>
+    /// The name of this rule - usefule for debugging
+    /// </summary>
+    public required string Name { get; init; }
+
+    /// <summary>
+    /// A list of node kinds to match. 
+    /// </summary>
+    public List<NodeKind> NodeKindList { get; set; } = new List<NodeKind>();
+
+    /// <summary>
+    /// A convenience accessor to match on a single NodeKind
+    /// </summary>
+    public NodeKind NodeKind {
+        get {
+            if (NodeKindList.Count != 1) {
+                throw new InvalidOperationException($"NodeKind is invalid when NodeKindList has {NodeKindList.Count} entries.");
+            }
+            return NodeKindList[0];
+        }
+        set {
+            NodeKindList = new List<NodeKind> { value };
+        }
+    }
+
+    /// <summary>
+    /// If and how to match the content of the Node
+    /// </summary>
+    public MatchNodeContentType MatchContent { get; init; } = MatchNodeContentType.None;
+
+    /// <summary>
+    /// Rules to apply to children node.
+    /// </summary>
+    public List<MatchRule> ChildRules { get; init; } = new List<MatchRule>();
+
+    /// <summary>
+    /// Whether or not the total number of children should match
+    /// </summary>
+    public bool MatchChildCount { get; init; } = false;
+
+    /// <summary>
+    /// A reducer to apply if the node matches
+    /// </summary>
+    public required Func<Node, string?, IReduction> Reducer { get; init; }
+
+    /// <summary>
+    /// Returns true if and only if the given node matches this rule
+    /// </summary>
+    /// <param name="n">a node to match on</param>
+    /// <returns></returns>
+    public bool Matches(Node n)
+    {
+        return NodeKindMatches(n) && ContentMatches(n) && ChildrenMatches(n);
+    }
+
+    /// <summary>
+    /// Returns true if and only if the NodeKind matches
+    /// </summary>
+    /// <param name="n">a node to match on</param>
+    /// <returns></returns>
+    bool NodeKindMatches(Node n)
+    {
+        return NodeKindList.Contains(n.Kind);
+    }
+
+    /// <summary>
+    /// Returns true if and only if the content of the given node matches
+    /// </summary>
+    /// <param name="n">a node to match on</param>
+    /// <returns></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    bool ContentMatches(Node n)
+    {
+        // Only care about the content type not its value
```

Comment:
Shouldn't we rename it to `ContentTypeMatches` if we are only care about the content type?